Scope pybind11 functions to torch.distributed.{autograd,rpc}

pietern opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27531 Run clang-format for torch/distributed/rpc
* #27530 Rename PythonUDF{Call,Resp}
* **#27529 Scope pybind11 functions to torch.distributed.{autograd,rpc}**

Differential Revision: [D17808209](https://our.internmc.facebook.com/intern/diff/D17808209)

pietern committed:
Scope pybind11 functions to torch.distributed.{autograd,rpc}

pietern committed:
Update on "Scope pybind11 functions to torch.distributed.{autograd,rpc}"

mrshenli commented on torch/distributed/rpc/__init__.py:
```diff
@@ -1,11 +1,21 @@
 
 
-if sys.version_info >= (3, 0):
+def is_available():
```
Do we then require `torch.distributed` to be available at least when constructing `ProcessGroupAgent`?

pietern replied:
We don't at the moment, but we should... I think it's safe to assume this is OK, though, because the RPC module is built for a subset of configurations where the distributed module is built.

mrshenli commented:
Test failure is real:

```
Oct 08 10:55:52 Running test_dist_autograd_fork ... [2019-10-08 10:55:52.553930]
Oct 08 10:55:52 Traceback (most recent call last):
Oct 08 10:55:52   File "test_dist_autograd_fork.py", line 4, in <module>
Oct 08 10:55:52     from dist_autograd_test import DistAutogradTest
Oct 08 10:55:52   File "/var/lib/jenkins/workspace/test/dist_autograd_test.py", line 9, in <module>
Oct 08 10:55:52     from dist_utils import INIT_METHOD_TEMPLATE, dist_init
Oct 08 10:55:52   File "/var/lib/jenkins/workspace/test/dist_utils.py", line 8, in <module>
Oct 08 10:55:52     from torch.distributed.rpc.api import RpcBackend
Oct 08 10:55:52   File "/opt/python/2.7.9/lib/python2.7/site-packages/torch/distributed/rpc/api.py", line 1, in <module>
Oct 08 10:55:52     from . import invoke_rpc_builtin, invoke_rpc_python_udf
Oct 08 10:55:52 ImportError: cannot import name invoke_rpc_builtin
Oct 08 10:55:52 Traceback (most recent call last):
Oct 08 10:55:52   File "test/run_test.py", line 458, in <module>
Oct 08 10:55:52     main()
Oct 08 10:55:52   File "test/run_test.py", line 450, in main
Oct 08 10:55:52     raise RuntimeError(message)
Oct 08 10:55:52 RuntimeError: test_dist_autograd_fork failed!
```

pietern commented:
The failure is real, but the problem is that `test_dist_autograd_fork` shouldn't even run for Python 2.

The imports it does are assumed to work, but this commit makes initialization of the RPC module (and the pybind11 functions) conditional on using Python 3.

pietern committed:
Update on "Scope pybind11 functions to torch.distributed.{autograd,rpc}"


Differential Revision: [D17808209](https://our.internmc.facebook.com/intern/diff/D17808209)

pietern committed:
Update on "Scope pybind11 functions to torch.distributed.{autograd,rpc}"


Differential Revision: [D17808209](https://our.internmc.facebook.com/intern/diff/D17808209)

pietern committed:
Update on "Scope pybind11 functions to torch.distributed.{autograd,rpc}"


Differential Revision: [D17808209](https://our.internmc.facebook.com/intern/diff/D17808209)

pietern commented:
The commit that's landing has a fix for the build failures in the most recent commit in this PR.
